### ISSUE TITLE
`explicit_counter_loop` suggests .into_iter(), despite that triggering `into_iter_on_ref` in some cases

### ISSUE THREAD

**Issue by 5225225** [rust-lang/rust-clippy#8155]
### Summary

I'm not sure if this is worth doing in the general case, but if you manually have a loop counter over an iterator over a slice, you'll be told to use `slice.into_iter().enumerate()`, but following that and re-running clippy will have clippy tell you to turn that `into_iter` into an `iter`.

### Reproducer

I tried this code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=5138a2b877837d0f09d4b2d4dcd2b2e0

```rust
#![allow(unused_variables)]

pub fn foo() {
    let slice = &[1, 2, 3];
    let mut i = 0;
    for item in slice {
        i += 1;
    }
}
```

I expected to see this happen:

```
    Checking playground v0.0.1 (/playground)
warning: the variable `i` is used as a loop counter
 --> src/lib.rs:6:5
  |
6 |     for item in slice {
  |     ^^^^^^^^^^^^^^^^^ help: consider using: `for (i, item) in slice.iter().enumerate()`
  |
  = note: `#[warn(clippy::explicit_counter_loop)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_counter_loop

warning: `playground` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.99s
```

Instead, this happened:

(the error as above, but with `into_iter` instead of `iter`)

### Version

```text
rustc 1.59.0-nightly (91a0600a5 2021-12-18)
binary: rustc
commit-hash: 91a0600a5c22b9d159e3c57526af83e71d1120f8
commit-date: 2021-12-18
host: x86_64-unknown-linux-gnu
release: 1.59.0-nightly
LLVM version: 13.0.0
```
